### DISCUSSION TITLE
v0.2.2 - Optimisations and stabilisation

### DISCUSSION THREAD

**Discussion by c3d** [c3d/db48x#75] [Announcements]
## Bugs fixed since v0.2.1:

- Fixed the `while` loop
  The test for the `while` loop was inverted
  The location of the `end` and `repeat` keywords was swapped

- Fixed bug in detection of multiplication overflow
  The computation of the factorial of 23 and up is now correct.

- Fixed bug mistakenly converting some fractions to integers
  For example, `2/17 INV` would return `17` instead of `17/2`

- Add spacing to large numbers
  A number that did not fit on 64-bit would not respect spacing settings

- Repaired the test suite
  Tests were setting-dependent, and would fail with default settings because
  they did not take number spacing into account.

## Features implemented since v0.2.1:

- Implemented local variables, e.g. `<< -> A B C << A B - C * >> >>`
  Local variables are a very fast way to reference data. Internally, they are
  represented by an offset to a known location, which makes them practically as
  fast as stack accesses, and much faster than global variable lookups.

- Implement the `same` function
  The `same` function detects if two objects are identical.
  Like the HP48, there are multiple "equality" variants:
  + `same` detects if two objects are binary-identical
  + `==` does the same, but dereference names
  + `=` is a mathematical equality (does not exist in HP48 RPL)

- Add feature to import ASCII versions of `«`, `»` and `→`
  In .48S files, you can use `<<`, `>>` and `->` instead.

- Implement `→List` (stack to list)
  This was the last missing command to be able to execute the
  [HP calculator's N-queens benchmark][benchmark]

[benchmark]: https://www.hpmuseum.org/cgi-sys/cgiwrap/hpmuseum/articles.cgi?read=700

- Better truncation of very large numbers when shown on stack
  A very large number, like the factorial of 64, would show up truncated
  at 80 digits, so the first digits shown were wrong.

- Various optimizations to make things a bit snappier, notably on battery power

  + Refactored the object model to use a simpler lookup mechanism

  + Implemented a C++ interface for arithmetic and functions, which will be
    useful when implementing complex number and matrix operations, and already
    simplified the code for operations such as `abs` or `cubed`.

  + Sime fine-tuning of LEB128 handling for types

  + Execution time for `NQueens` measured, reduced from about 980ms down
    to about 530ms.

**Full Changelog**: https://github.com/c3d/DB48X-on-DM42/compare/v0.2.1...v0.2.2

<hr /><em>This discussion was created from the release <a href='https://github.com/c3d/DB48X-on-DM42/releases/tag/v0.2.2'>v0.2.2 - Optimisations and stabilisation</a>.</em>

**Comment by balta1980** [comment]
Already on my calc, and testing.

**Reply by balta1980** [reply]
After using a while the calc I found that:
All on battery:
1) fixed modes are not working, perhaps pressing 2 FIX, and operating 0.333333 * 1 result is 0.333333

2) when entering an rpl routine the calc crash randomly, but very often. It also crash after you can evaluate the routine.

3) thousands separator are not working, no matter what setting is set.

**Reply by balta1980** [reply]
Same behavior when plugged to USB

**Comment by c3d** [comment]
1. is a regression. I filed #100, I'll look into it.
2. I noticed a similar problem while testing `for` loops, filed as #101.
3. This is probably the same as 1. I just fixed it while working on your 17B suggestion.

**Reply by c3d** [reply]
@balta1980 Would you mind checking that these problems are gone with [v0.2.4](https://github.com/c3d/DB48X-on-DM42/releases/tag/v0.2.4)? Thanks!

**Reply by balta1980** [reply]
I can confirm that on 0.2.4 these problems are gone on battery and USB plugged. Thank you.

**Comment by c3d** [comment]
@balta1980 Note that I already found a couple of issues with v0.2.4 (already fixed on my dev branch):
- The CHS key no longer works correctly on the second number on the command line (or exponent of first number).
- Program interruption with EXIT does not work if you typed other keys before it.
